src/internal/curry.mjs

```diff
@@ -1,10 +1,20 @@
 export function curry (fn, expectedArgsLength = fn.length, appliedArgs = []) {
-  return (...args) => {
+  function curryInternal (...args) {
     if (expectedArgsLength <= args.length) {
       return fn(...appliedArgs, ...args)
     }
-    return curry(fn, expectedArgsLength - args.length, [...appliedArgs, ...args])
+    const argsLength = args.length
+    args.unshift(...appliedArgs)
+    return curry(fn, expectedArgsLength - argsLength, args)
   }
+
+  curryInternal.partial = function (...args) {
```

Comment:
Dot pattern looks pretty, but it may prevent tree-shaking. Why don't you use `partialCurryInternal` instead?

Reply:
You are in fact both right and wrong. The . pattern does not prevent tree shaking. You can see that clearly [here](https://rollupjs.org/repl?version=1.2.3&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QnNxdWFyZSUyQyUyMGN1YmUlN0QlMjBmcm9tJTIwJy4lMkZtYXRocy5qcyclNUNuJTVDbmNvbnNvbGUubG9nKHNxdWFyZSg1KSklM0IlNUNuJTJGJTJGY29uc29sZS5sb2coY3ViZSg1KSklM0IlMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJtYXRocy5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJpbXBvcnQlMjBzcXVhcmUlMjBmcm9tJTIwJy4lMkZzcXVhcmUuanMnJTNCJTVDbiU1Q24lMjBmdW5jdGlvbiUyMGN1YmUlMjAoeCklMjAlN0IlNUNuJTVDdHJldHVybiUyMHNxdWFyZSh4KSUyMColMjB4JTNCJTVDbiU3RCU1Q24lNUNuY3ViZS5mb28lMjAlM0QlMjBmdW5jdGlvbigpJTIwJTdCJTdEJTVDbiU1Q24lMjBleHBvcnQlMjAlN0IlMjBjdWJlJTJDJTIwc3F1YXJlJTIwJTdEJTIyJTJDJTIyaXNFbnRyeSUyMiUzQWZhbHNlJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMnNxdWFyZS5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJleHBvcnQlMjBkZWZhdWx0JTIwZnVuY3Rpb24lMjBzcXVhcmUlMjAoJTIweCUyMCklMjAlN0IlNUNuJTVDdHJldHVybiUyMHglMjAqJTIweCUzQiU1Q24lN0QlMjIlMkMlMjJpc0VudHJ5JTIyJTNBZmFsc2UlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyY2pzJTIyJTJDJTIybmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=)

But also, the curry function itself does very much prevent all tree shaking. We need to go through and mark every call site of `iterableCurry` and `asyncIterableCurry` with `#__PURE__ ` comments.

Reply:
Also adding the pure comments will ensure that it doesn't matter what happens inside curry.

Reply:
In [your example][conartist example], `cube.foo` is defined in global context, thus tree-shakable. But your code is more akin to [this][my example] (every method is defined in local context, just like `partial`).


[conartist example]: https://rollupjs.org/repl?version=1.2.3&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QnNxdWFyZSUyQyUyMGN1YmUlN0QlMjBmcm9tJTIwJy4lMkZtYXRocy5qcyclNUNuJTVDbmNvbnNvbGUubG9nKHNxdWFyZSg1KSklM0IlNUNuJTJGJTJGY29uc29sZS5sb2coY3ViZSg1KSklM0IlMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJtYXRocy5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJpbXBvcnQlMjBzcXVhcmUlMjBmcm9tJTIwJy4lMkZzcXVhcmUuanMnJTNCJTVDbiU1Q24lMjBmdW5jdGlvbiUyMGN1YmUlMjAoeCklMjAlN0IlNUNuJTVDdHJldHVybiUyMHNxdWFyZSh4KSUyMColMjB4JTNCJTVDbiU3RCU1Q24lNUNuY3ViZS5mb28lMjAlM0QlMjBmdW5jdGlvbigpJTIwJTdCJTdEJTVDbiU1Q24lMjBleHBvcnQlMjAlN0IlMjBjdWJlJTJDJTIwc3F1YXJlJTIwJTdEJTIyJTJDJTIyaXNFbnRyeSUyMiUzQWZhbHNlJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMnNxdWFyZS5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJleHBvcnQlMjBkZWZhdWx0JTIwZnVuY3Rpb24lMjBzcXVhcmUlMjAoJTIweCUyMCklMjAlN0IlNUNuJTVDdHJldHVybiUyMHglMjAqJTIweCUzQiU1Q24lN0QlMjIlMkMlMjJpc0VudHJ5JTIyJTNBZmFsc2UlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyY2pzJTIyJTJDJTIybmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=

[my example]: https://rollupjs.org/repl?version=1.2.3&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMGNyZWF0ZUZ1bmN0aW9ucyUyMGZyb20lMjAnLiUyRmZhY3RvcnkuanMnJTVDbiU1Q25jb25zdCUyMCU3QiUyMGZvbyUyMCU3RCUyMCUzRCUyMGNyZWF0ZUZ1bmN0aW9ucygpJTVDbiU1Q25jb25zb2xlLmxvZyhmb28oKSklMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJmYWN0b3J5LmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGRlZmF1bHQlMjBmdW5jdGlvbiUyMGNyZWF0ZUZ1bmN0aW9ucyUyMCgpJTIwJTdCJTVDbiU1Q3Rjb25zdCUyMHJlc3VsdCUyMCUzRCUyMCU3QiU3RCU1Q24lNUN0cmVzdWx0LmZvbyUyMCUzRCUyMCgpJTIwJTNEJTNFJTIwJ2ZvbyclNUNuJTVDdHJlc3VsdC5iYXIlMjAlM0QlMjAoKSUyMCUzRCUzRSUyMCdiYXInJTVDbiU1Q3RyZXN1bHQuYmF6JTIwJTNEJTIwKCklMjAlM0QlM0UlMjAnYmF6JyU1Q24lNUN0cmV0dXJuJTIwcmVzdWx0JTVDbiU3RCUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJjanMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==